src/probability/upcrossing.lean

```diff
@@ -0,0 +1,914 @@
+/-
+Copyright (c) 2022 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import probability.hitting_time
+import probability.martingale
+
+/-!
+
+# Doob's upcrossing estimate
+
+Given a discrete real-valued submartingale $(f_n)_{n \in \mathbb{N}}$, denoting $U_N(a, b)$ the
+number of times $f_n$ crossed from below $a$ to above $b$ before time $N$, Doob's upcrossing
+estimate (also known as Doob's inequality) states that
+$$(b - a) \mathbb{E}[U_N(a, b)] \le \mathbb{E}[(f_N - a)^+].$$
+Doob's upcrossing estimate is an important inequality and is central in proving the martingale
+convergence theorems.
+
+## Main definitions
+
+* `measure_theory.upper_crossing_time a b f N n`: is the stopping time corresponding to `f`
+  crossing above `b` the `n`-th time before time `N` (if this does not occur then the value is
+  taken to be `N`).
+* `measure_theory.lower_crossing_time a b f N n`: is the stopping time corresponding to `f`
+  crossing below `a` the `n`-th time before time `N` (if this does not occur then the value is
+  taken to be `N`).
+* `measure_theory.upcrossing_strat a b f N`: is the predictable process which is 1 if `n` is
+  between a consecutive pair of lower and upper crossing and is 0 otherwise. Intuitively
+  one might think of the `upcrossing_strat` as the strategy of buying 1 share whenever the process
+  crosses below `a` for the first time after selling and selling 1 share whenever the process
+  crosses above `b` for the first time after buying.
+* `measure_theory.upcrossings_before a b f N`: is the number of times `f` crosses from below `a` to
+  above `b` before time `N`.
+* `measure_theory.upcrossings a b f`: is the number of times `f` crosses from below `a` to above
+  `b`. This takes value in `ℝ≥0∞` and so is allowed to be `∞`.
+
+## Main results
+
+* `measure_theory.adapted.is_stopping_time_upper_crossing_time`: `upper_crossing_time` is a
+  stopping time whenever the process it is associated to is adapted.
+* `measure_theory.adapted.is_stopping_time_lower_crossing_time`: `lower_crossing_time` is a
+  stopping time whenever the process it is associated to is adapted.
+* `measure_theory.submartingale.mul_integral_upcrossings_before_le_integral_pos_part`: Doob's
+  upcrossing estimate.
+* `measure_theory.submartingale.mul_lintegral_upcrossing_le_lintegral_pos_part`: the inequality
+  obtained by taking the supremum on both sides of Doob's upcrossing estimate.
+
+### References
+
+We mostly follow the proof from [Kallenberg, *Foundations of modern probability*][kallenberg2021]
+
+-/
+
+open topological_space filter
+open_locale nnreal ennreal measure_theory probability_theory big_operators topological_space
+
+namespace measure_theory
+
+variables {α ι : Type*} {m0 : measurable_space α} {μ : measure α}
```

Comment:
I've changed `α` to `Ω` and `x` to `ω`. I've not switched `μ` to `ℙ` since this file had opened `probability.notation` and it seems the notation of `ℙ` being the volume doesn't allow me to declare it as a measure (I can't use `measure_space` here since I want to specify the `measurable_space`).